### PR TITLE
Support `no_std` and `alloc` builds for the protocol parser

### LINKED ISSUE #70: Feature request: `no_std + alloc` parser
The parser in `src/protocol.rs` is pure logic but pulls `std::collections::HashMap`, `std::sync::Arc`, `std::time::SystemTime`, blocking adoption by embedded and wasm consumers.

I propose to add `#![cfg_attr(not(feature = "std"), no_std)]` on `lib.rs`, swap `std::` for `alloc::` plus `hashbrown` and `smallvec`, defaults unchanged. Connection backends (`libpq`, `rustls-tls`) stay `std`-only. Pairs with the optional-`tokio` issue.

I am working on this now. Filing first to align on the feature name (`std` default-on vs `alloc` opt-in) and the `SystemTime` substitution (`chrono::DateTime<Utc>` is the obvious move) before the PR.

### PR DESCRIPTION
**Pull request by LucaCappelletti94**
Makes the protocol parser build in `no_std` plus `alloc` (wasm and embedded) with the default build untouched: `std` is a default feature, `--no-default-features` gives the parser-only build, and the `libpq` and `rustls-tls` backends pull `std`. Builds on the optional `tokio` support from #69.

`RelationMap` and `ChangeEvent.metadata` move to `alloc::collections::BTreeMap` rather than `hashbrown` (no extra crate, benchmarked within noise of the old `HashMap`), and `chrono::DateTime<Utc>` stays the timestamp type with `chrono` at `default-features = false` since the parser builds it from the wire `i64`, never a clock. `SystemTime`, the `Instant` feedback timing, and the `Io`/`StringConversion` error variants are `std`-gated, and CI adds a `wasm32-unknown-unknown` build.

Closes #70.

### CODE REVIEW

File: src/buffer.rs [hunk 1]
```diff
@@ -163,17 +164,17 @@ impl BufferReader {
 
     /// Read a null-terminated string directly into `Arc<str>` (avoids intermediate String allocation).
     #[inline]
-    pub fn read_cstring_arc(&mut self) -> Result<std::sync::Arc<str>> {
+    pub fn read_cstring_arc(&mut self) -> Result<alloc::sync::Arc<str>> {
```
[the remark this answers is not in the document]
Comment (LucaCappelletti94) [reply src/buffer.rs]: Fixed

File: src/buffer.rs [hunk 2]
```diff
@@ -163,17 +164,17 @@ impl BufferReader {
...
         let bytes_to_read = memchr::memchr(0, data_slice).ok_or_else(|| {
             ReplicationError::protocol("Unterminated string in buffer".to_string())
         })?;
 
-        let result = std::str::from_utf8(&data_slice[..bytes_to_read])
+        let result = core::str::from_utf8(&data_slice[..bytes_to_read])
             .map_err(|e| ReplicationError::protocol(format!("Invalid UTF-8 in string: {e}")))?;
 
-        let arc: std::sync::Arc<str> = std::sync::Arc::from(result);
+        let arc: alloc::sync::Arc<str> = alloc::sync::Arc::from(result);
```
[the remark this answers is not in the document]
Comment (LucaCappelletti94) [reply src/buffer.rs]: Fixed

File: src/deserializer.rs [hunk 3]
```diff
@@ -127,7 +127,7 @@ where
 
 /// Render a byte slice as a string for error messages, lossy on non-UTF-8.
 #[inline]
-fn lossy_token(b: &[u8]) -> std::borrow::Cow<'_, str> {
+fn lossy_token(b: &[u8]) -> alloc::borrow::Cow<'_, str> {
```
[the remark this answers is not in the document]
Comment (LucaCappelletti94) [reply src/deserializer.rs]: Fixed

File: src/types.rs [hunk 4]
```diff
@@ -2513,7 +2514,7 @@ mod tests {
         let decoded = ChangeEvent::decode(&buf).expect("decode failed");
         assert_eq!(decoded.lsn, event.lsn);
         assert_eq!(decoded.event_type, event.event_type);
-        // Compare metadata (HashMap doesn't impl Eq but we can check the content)
+        // Compare metadata (BTreeMap doesn't impl Eq but we can check the content)
```
[the remark this answers is not in the document]
Comment (LucaCappelletti94) [reply src/types.rs]: Fixed

**Comment by isdaniel** [comment]
Hi @LucaCappelletti94 thank you for your effort, I have two comments after I reviews your PR, please kindly check.

1. add CIDC testing for no_std aarch64-unknown-none    

like:

- name: Build no_std (alloc) for aarch64-unknown-none (bare-metal)                  
  run: cargo build --no-default-features --lib --target aarch64-unknown-none                                                                                

2. split LSN-recording from the std-only feedback timestamp, and change code accordlty

`lsn_has_changed()` is callable under no_std, but the only setter for the last_sent_* baseline it compares against  `mark_feedback_sent_with_lsn()`  is #[cfg(feature = "std")] (it calls Instant::now()). 

So a no_std consumer can read via lsn_has_changed() but can never advance the baseline: it stays stuck comparing against the initial 0 and always returns true, making the dedup API effectively dead off-std.

File: src/protocol.rs [hunk 5]
```diff
@@ -669,13 +638,15 @@ impl ReplicationState {
         flush_lsn != self.last_sent_flush_lsn || applied_lsn != self.last_sent_applied_lsn
     }
 
+    #[cfg(feature = "std")]
     /// Mark feedback as sent and record the LSN values sent
     pub fn mark_feedback_sent_with_lsn(&mut self, flush_lsn: XLogRecPtr, applied_lsn: XLogRecPtr) {
```
Comment (isdaniel) [inline src/protocol.rs:651]: lsn_has_changed() is callable under no_std, but the only setter for the last_sent_* baseline it compares against mark_feedback_sent_with_lsn() is #[cfg(feature = "std")] (it calls Instant::now()).

Comment (LucaCappelletti94) [reply src/protocol.rs:651]: The whole feedback-send path is std-only, since it needs both the connection and a clock, so I gated the rest of it to match. I had proceeded mechanically in the replacement and had not noticed this method should arguably also be gated. The reader and its only writer now sit on the same side. In `no_std` the `ReplicationState` keeps just the relation cache plus the raw `last_received`/`last_flushed`/`last_applied` counters and their `update_*` setters. Done in the follow-up commit `17b49cd`.

**Comment by isdaniel** [comment]
Hi @LucaCappelletti94 thank you for your modifications. look good to me now, could you kindly rebase your commits let it more clear? after that please let me know i will do final check before merged this PR.

**Comment by LucaCappelletti94** [comment]
Rebased into a single commit on the latest `main`. Now:

1. Bare-metal CI: the `no_std` job now also runs `cargo build --no-default-features --lib --target aarch64-unknown-none`. `wasm32-unknown-unknown` ships a partial `std` that could quietly mask an accidental `std` dependency, whereas `aarch64-unknown-none` has no `std` at all, so it gates the whole dependency graph. It builds clean.

2. LSN split: `lsn_has_changed` and the `last_sent_*` baseline are available in `no_std` again, and `mark_feedback_sent_with_lsn` now sets the `Instant` timestamp and delegates the recording to a new `record_sent_lsns(flush, applied)` that is `no_std`-callable. So an off-`std` consumer can dedup with `lsn_has_changed` and advance the baseline with `record_sent_lsns`, with only the clock staying `std`.

My first attempt at the feedback comment went the wrong direction. I gated `lsn_has_changed` and the baseline out of `no_std` entirely, which removed the dedup off-`std` instead of making it usable, as I had misread your request.

**Comment by isdaniel** [comment]
LGTM, now, thank you for your contribution
